### PR TITLE
[Ven 1463] [VEN-1464]: Abstract swapper methods(1/4)

### PR DESCRIPTION
**Pull request by Debugger022**
Resolves:
- VEN-1463
- VEN-1464

### CODE REVIEW

File: contracts/PassiveSwap/AbstractSwapper.sol [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+contract AbstractSwapper is AccessControlledV8 {
+    /// @notice Swap configuration for the tokens pair
+    struct SwapConfiguration {
+        /// tokenIn address
+        address tokenAddressIn;
+        /// tokenOut address
+        address tokenAddressOut;
+        /// incentive on swapping tokens
+        uint256 incentive;
```
Comment (chechu) [inline contracts/PassiveSwap/AbstractSwapper.sol:18]: I would add a comment to clarify the unit of this variable, and I would give some examples. AFAIU, `10 * 1e18` would be an incentive of 10%, right?

File: contracts/PassiveSwap/AbstractSwapper.sol [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        /// whether the swap is enabled
+        bool enabled;
+    }
+
+    ResilientOracle public priceOracle;
+
```
Comment (GitGuru7) [inline contracts/PassiveSwap/AbstractSwapper.sol:24]: This contract will be inherited in child contract for that reason we should add a `--gap[]` variable here

File: contracts/PassiveSwap/AbstractSwapper.sol [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    /// @notice swap configurations for the existing pairs
+    mapping(address => mapping(address => SwapConfiguration)) public swapConfigurations;
+
+    /// @notice Emitted when config is updated for tokens pair
+    event SwapConfigurationUpdated(
+        address tokenAddressIn,
+        address tokenAddressOut,
```
[the remark this answers is not in the document]
Comment (Debugger022) [reply contracts/PassiveSwap/AbstractSwapper.sol]: Added, d1c6ee3894085fd33e8ac98110d52004eab99ac9

File: contracts/PassiveSwap/AbstractSwapper.sol [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        address tokenAddressOut;
+        /// incentive on swapping tokens
+        uint256 incentive;
+        /// whether the swap is enabled
+        bool enabled;
+    }
+
+    ResilientOracle public priceOracle;
+
+    /// @notice swap configurations for the existing pairs
```
[the remark this answers is not in the document]
Comment (Debugger022) [reply contracts/PassiveSwap/AbstractSwapper.sol:30]: Added, d1c6ee3894085fd33e8ac98110d52004eab99ac9

File: contracts/PassiveSwap/AbstractSwapper.sol [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            revert SwapConfigurationNotEnabled();
+        }
+
+        uint256 maxTokenOutLiquidity = VTokenInterface(tokenAddressOut).balanceOf(address(this));
+        uint256 tokenInUnderlyingPrice = priceOracle.getUnderlyingPrice(tokenAddressIn);
+        uint256 tokenOutUnderlyingPrice = priceOracle.getUnderlyingPrice(tokenAddressOut);
+
+        if (tokenInUnderlyingPrice == 0 || tokenOutUnderlyingPrice == 0) {
+            revert InvalidOraclePrice();
+        }
```
[the remark this answers is not in the document]
Comment (Debugger022) [reply contracts/PassiveSwap/AbstractSwapper.sol]: Removed check and event, d1c6ee3894085fd33e8ac98110d52004eab99ac9

File: contracts/PassiveSwap/AbstractSwapper.sol [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            revert InsufficientInputAmount();
+        }
+
+        SwapConfiguration storage configuration = swapConfigurations[tokenAddressIn][tokenAddressOut];
+
+        if (!configuration.enabled) {
+            revert SwapConfigurationNotEnabled();
+        }
+
+        uint256 maxTokenOutLiquidity = VTokenInterface(tokenAddressOut).balanceOf(address(this));
```
[the remark this answers is not in the document]
Comment (Debugger022) [reply contracts/PassiveSwap/AbstractSwapper.sol:121]: RN oracle PR is not merged so we can use it, once it gets merged then we can refactor here

File: contracts/PassiveSwap/AbstractSwapper.sol (hunk 1 above)
Comment (Debugger022) [reply contracts/PassiveSwap/AbstractSwapper.sol]: No, 10% incentive means 0.1 * 1e18
I have added comment also, d1c6ee3894085fd33e8ac98110d52004eab99ac9

File: contracts/PassiveSwap/AbstractSwapper.sol (hunk 2 above)
Comment (Debugger022) [reply contracts/PassiveSwap/AbstractSwapper.sol:24]: Added, d1c6ee3894085fd33e8ac98110d52004eab99ac9

File: contracts/PassiveSwap/AbstractSwapper.sol [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+    /// @notice Thrown when underlying price is zero
+    error InvalidOraclePrice();
+
+    /**
+     * @param accessControlManager_ Access control manager contract address
+     */
+    function initialize(address accessControlManager_, ResilientOracle priceOracle_) external initializer {
+        __AccessControlled_init(accessControlManager_);
+        priceOracle = priceOracle_;
```
[the remark this answers is not in the document]
Comment (Debugger022) [reply contracts/PassiveSwap/AbstractSwapper.sol]: I have added internal as well as external setter for the priceOracle along event and unit tests, c566a1e843dc48ee7f6702683735ca0aaf6c5bd4

[the remark this answers is not in the document]
Comment (Debugger022) [reply contracts/PassiveSwap/AbstractSwapper.sol]: Added a constant `MAX_INCENTIVE` and it's value is `5e18`. Added check for the incentive while updating for a pair.
f9e28c8f34751b2e17222544721de12e272a1de8

File: contracts/PassiveSwap/AbstractSwapper.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (Debugger022) [reply contracts/PassiveSwap/AbstractSwapper.sol:32]: Yes, we can do that after completing the development in separate PR.

File: contracts/PassiveSwap/AbstractSwapper.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (Debugger022) [reply contracts/PassiveSwap/AbstractSwapper.sol:130]: Yes, I have updated it in another PR, c97b204ec21c32420f9b2f6e847073c8165767eb

**Review by GitGuru7 — APPROVED** [review]

**Pull request merged**